Log a trace event on process termination

ajbeamon opened the pull request:
This also adds SIGQUIT to the crash handler.

I haven't yet carefully looked at how signal-safe this is. We do something similar in the crash handler (in addition to other things), but in that case we deemed it acceptable because we were already crashing.

sfc-gh-tclinkenbeard commented on flow/Platform.cpp:
```diff
@@ -3171,13 +3171,26 @@ void platformInit() {
 	std::string backtrace = platform::get_backtrace();
 
-	bool error = (sig != SIGUSR2);
+	bool error = (sig != SIGUSR2 || sig != SIGQUIT);
```
```suggestion
	bool error = (sig != SIGUSR2 && sig != SIGQUIT);
```

ajbeamon commented:
It looks like the work to log and flush the trace events requires a mutex, which would be a problem.

ajbeamon commented:
I've updated this to instead stop the network thread upon receiving a termination signal, and then also added a trace event to be logged when the network thread closes.

I'm not certain whether this is signal safe as written, but it's an approach that should be possible to make signal safe.

We should discuss whether we are happy handling the signals in this way, as it could mean that the process fails to die when terminated in some cases (though a kill -9 should still work as it does today).

We could also potentially record some additional info to be logged in the termination event, such as the fact that we were terminated by signal and what the signal was.

sfc-gh-ajbeamon commented:
This was reopened as #4477, so this one can be closed.
